cpp/src/arrow/compute/kernels/cast.cc

```diff
@@ -660,6 +663,100 @@ struct CastFunctor<T, DictionaryType,
   }
 };
 
+// ----------------------------------------------------------------------
+// String to Number
+
+template <typename O>
+struct CastFunctor<O, StringType,
+                   typename std::enable_if<std::is_base_of<Number, O>::value>::type> {
+  void operator()(FunctionContext* ctx, const CastOptions& options,
+                  const ArrayData& input, ArrayData* output) {
+    using out_type = typename O::c_type;
+    StringArray input_array(input.Copy());
+
+    if (input_array.null_count() > 0) {
+      std::stringstream ss;
+      ss << "Failed to cast NA into " << output->type->ToString();
+      ctx->SetStatus(Status(StatusCode::SerializationError, ss.str()));
+      return;
+    }
```

Comment:
If the input has nulls, then the output should have nulls in the same locations (like the other cast functions)